feat(avm): Constrain Poseidon2 Lookups by Round

MirandaWood opened the pull request:
Closes #12738

Adds columns to constrain that poseidon2 hash lookups include the rounds remaining, so we cannot misorder/insert rows to produce a proof with an incorrect hash. Note that the final row of `poseidon2_hash` constrains that `num_perm_rounds_rem = 1`, so some lookups which use `poseidon2_hash.end` did not need to be changed.

Hashes over a single row are generally constrained by looking up `poseidon2_hash.end = poseidon2_hash.start = 1`, since the trace will ensure we just have one row. The following circuits already did this and didn't need to be further constrained:

* protocol_contract.pil
* note_hash_tree_check.pil
* retrieved_bytecodes_tree_check.pil
* written_public_data_slots_tree_check.pil

MirandaWood commented on barretenberg/cpp/pil/vm2/bytecode/address_derivation.pil:
```diff
@@ -39,9 +39,19 @@ namespace address_derivation;
-    sel { partial_address_domain_separator, salt, init_hash, salted_init_hash }
-    in poseidon2_hash.start { poseidon2_hash.input_0, poseidon2_hash.input_1, poseidon2_hash.input_2, poseidon2_hash.output };
+    sel { partial_address_domain_separator, salt, init_hash, salted_init_hash, const_four }
+    in poseidon2_hash.start { poseidon2_hash.input_0, poseidon2_hash.input_1, poseidon2_hash.input_2, poseidon2_hash.output, poseidon2_hash.input_len };
```
To avoid defining yet another constant column, I'm looking up `input_len` rather than `rounds` - `input_len` is constrained at `start = 1` so I think this is fine!

jeanmon replied:
These ones require two rows, so why do not we use the same technique and using `const_two`?

MirandaWood replied:
Can be either! I just thought this was more readable. I can switch to `two` to keep it consistent with the other lookups though

MirandaWood commented on barretenberg/cpp/pil/vm2/trees/merkle_check.pil:
```diff
@@ -161,17 +161,14 @@ namespace merkle_check;
     pol commit read_output_hash;
     pol commit write_output_hash;
 
-    pol commit constant_2;
-    sel * (constant_2 - 2) = 0;
-
```
We can just use `start` and `end` for the lookups in this file since they are both single row hashes, meaning we don't need this column

jeanmon commented on barretenberg/cpp/pil/vm2/bytecode/address_derivation.pil:
```diff
@@ -39,9 +39,19 @@ namespace address_derivation;
     pol commit partial_address_domain_separator;
     sel * (partial_address_domain_separator - constants.GENERATOR_INDEX__PARTIAL_ADDRESS) = 0;
 
+    // TODO: We need these temporarily while we dont allow for aliases in the lookup tuple
```
Nothing to do with this PR, but by grepping the source code, I have found an include to "poseidon2_hash.pil" from l1_to_l2_message_tree_check.pil which is not used.
Would you mind removing it?

jeanmon commented on barretenberg/cpp/pil/vm2/bytecode/address_derivation.pil:
```diff
@@ -39,9 +39,19 @@ namespace address_derivation;
     sel * (partial_address_domain_separator - constants.GENERATOR_INDEX__PARTIAL_ADDRESS) = 0;
 
+    // TODO: We need these temporarily while we dont allow for aliases in the lookup tuple
+    pol commit const_two;
```
Add the standard TODO about removal of these constants once lookup support them.

MirandaWood replied:
I added one just above - do you mean paste it for each individual column?

jeanmon replied:
I think I was drunk.

jeanmon commented on barretenberg/cpp/pil/vm2/bytecode/bc_hashing.pil:
```diff
@@ -75,15 +75,16 @@ namespace bc_hashing;
     // Needs to be a committed column as it is used in the lookup:
     pol commit sel_not_start;
-    sel_not_start * (1 - sel_not_start) = 0;
+    sel_not_start - sel * (1 - start) = 0;
```
```suggestion
    sel_not_start = sel * (1 - start);
```

jeanmon commented on barretenberg/cpp/pil/vm2/bytecode/class_id_derivation.pil:
```diff
@@ -23,10 +23,12 @@ namespace class_id_derivation;
     // TODO: We need this temporarily while we dont allow for aliases in the lookup tuple, there must be a better way
     pol commit temp_constant_for_lookup;
     sel * (temp_constant_for_lookup - constants.GENERATOR_INDEX__CONTRACT_LEAF) = 0;
+    pol commit const_two;
```
Add the TODO about lookup supporting constants.

MirandaWood replied:
As above - the comment exists at the top (though here I will change it from `this` to `these`). I can copy and paste it one line below if required though

jeanmon replied:
My bad I hallucinated :-)

jeanmon reviewed (changes requested):
Great overall.
It sounds a bit subtle to correctly us poseidon2_hash..pil.
Could you please add a "usage" at the beginning of poseidon2_hash?
In particular, emphasizing the safe patterns:
1) If single row, embed `start` in the lookup and use destination selector `end`
2) For multiple rows, add the "remaining rounds" for every row except the last one. Use `start` selector in the first row and `end` selector in the last one.

MirandaWood committed:
feat(avm): Constrain Poseidon2 Lookups by Round

Closes #12738

Adds columns to constrain that poseidon2 hash lookups include the rounds remaining, so we cannot misorder/insert rows to produce a proof with an incorrect hash. Note that the final row of `poseidon2_hash` constrains that `num_perm_rounds_rem = 1`, so some lookups which use `poseidon2_hash.end` did not need to be changed.

Hashes over a single row are generally constrained by looking up `poseidon2_hash.end = poseidon2_hash.start = 1`, since the trace will ensure we just have one row. The following circuits already did this and didn't need to be further constrained:

* protocol_contract.pil
* note_hash_tree_check.pil
* retrieved_bytecodes_tree_check.pil
* written_public_data_slots_tree_check.pil
